gcc/rust/expand/rust-macro-builtins.cc

```diff
@@ -252,4 +263,42 @@ MacroBuiltin::compile_error (Location invoc_locus, AST::MacroInvocData &invoc)
   return AST::ASTFragment::create_error ();
 }
 
+/* Expand builtin macro concat!(), which joins all the literal parameters
+   into a string with no delimiter. */
+
+AST::ASTFragment
+MacroBuiltin::concat (Location invoc_locus, AST::MacroInvocData &invoc)
+{
+  auto invoc_token_tree = invoc.get_delim_tok_tree ();
+  MacroInvocLexer lex (invoc_token_tree.to_token_stream ());
+  Parser<MacroInvocLexer> parser (std::move (lex));
+  auto str = std::string ();
+
+  auto last_token_id = macro_end_token (invoc_token_tree, parser);
+
+  /* NOTE: concat! could accept no argument, so we don't have any checks here */
+  while (parser.peek_current_token ()->get_id () != last_token_id)
+    {
+      auto token = parser.peek_current_token ();
+      if (token->is_literal () || token->get_id () == TRUE_LITERAL
+	  || token->get_id () == FALSE_LITERAL)
+	{
+	  auto lit_expr = parser.parse_literal_expr ();
+	  str += lit_expr->as_string ();
+	}
+      else
+	{
+	  rust_error_at (parser.peek_current_token ()->get_locus (),
+			 "argument must be a constant literal");
+	  return AST::ASTFragment::create_error ();
+	}
+      parser.maybe_skip_token (COMMA);
```

Comment:
```suggestion
      auto literal = parser.parse_literal_expr();
      if (!literal) // or literal == nullptr
      {
          rust_error_at (parser.peek_current_token ()->get_locus (), "argument to %<concat%> must be a constant literal");
          return AST::ASTFragment::create_error ();
      }
     
      parser.maybe_skip_token (COMMA);
```
I think you might be able to simplify the logic here?

Reply:
Thank you. However, should we apply the following changes to `gcc/rust/lex/rust-token.h`?

```diff
diff --git a/gcc/rust/lex/rust-token.h b/gcc/rust/lex/rust-token.h
index bbc6d5ddf..3a7c63d4e 100644
--- a/gcc/rust/lex/rust-token.h
+++ b/gcc/rust/lex/rust-token.h
@@ -434,6 +434,8 @@ return *str;
       case STRING_LITERAL:
       case BYTE_CHAR_LITERAL:
       case BYTE_STRING_LITERAL:
+      case TRUE_LITERAL:
+      case FALSE_LITERAL:
        return true;
       default:
        return false;
```

Reply:
Yeah that could be pretty useful in other passes